### ISSUE TITLE
improper use of FORTRAN KIND # for COMPLEX variable declaration

### ISSUE THREAD

**Issue by richardbleikamp** [SOLLVE/sollve_vv#41]
**Describe the bug**
One of the sollve_vv tests contains the following declaration:
>        COMPLEX(kind = 16) :: scalar_double_complex
twice.
This declaration was intended to make "scalar_double_complex" a complex variable whose
real and imaginary parts are DOUBLE PRECISION (in fortran terms).
However, the KIND=n syntax introduced in Fortran 90 does NOT imitate the older non-standard
syntax every compiler supports, namely  "COMPLEX*16 scalar_double_complex".
The non-standard syntax specifies a byte count for the entire variable.  Hence,
>      COMPLEX*16 z
declares "z" to be two consecutive 8 bytes real types (the real and imaginary parts of "z"), while
the newer
>      COMPLEX (KIND=16) z
makes both the real and imaginary parts of "z" and REAL*16 type, and the entire variable is 32 bytes long.

So, the declaration should be either
>            COMPLEX(kind = 8) :: scalar_double_complex
or (less desirabe)
>            COMPLEX*16 scalar_double_complex

**Test that it applies to**
> tests/4.5/target/test_target_defaultmap.F90

**To Reproduce**
n/a

**Expected behavior**
After the change, a compiler like flang, that does not support REAL*16 type, will be able to compile the code.

**Compiler**
Any.

**Accelerator hardware**
n/a

**Comment by josemonsalve2** [comment]
Thanks @richardbleikamp!

**Comment by dmcdougall** [comment]
Ahhh!  I have been wondering about this.  This may fix a compiler error I get with `gfortran` too...  Let me try it and report back.

**Comment by frederik-h** [comment]
> Ahhh! I have been wondering about this. This may fix a compiler error I get with `gfortran` too... Let me try it and report back.

The change suggested by @richardbleikamp (both variants) makes the test compile with the current GCC trunk. However, for me, execution fails both on the host as well as on an offloading device (AMD GCN) whereas the corresponding C test passes with both setups. But the failure does not seem to be related to the change.

**Comment by josemonsalve2** [comment]
@frederik-h does it mean you suggest a different route? Or do you mean this is still a problem that GCC has to work on?

**Comment by dmcdougall** [comment]
Ok the change fixes the compile.  Here's the runtime error I'm getting:

```
 running: bin/test_target_defaultmap.F90.FOR.run 
/data/damon/compiler_tests/sollve_vv/sys/scripts/run_test.sh bin/test_target_defaultmap.F90.o 1 2>&1 | tee -a logs/test_target_defaultmap.F90.log && echo "PASS" > _ompvv_temp_result_.exitstatus || echo "FAIL" >
_ompvv_temp_result_.exitstatus
test_target_defaultmap.F90.o: FAIL. exit code: 1
test_target_defaultmap.F90.o:
[OMPVV_INFOMSG test_target_defaultmap.F90:17] Test is running on device
[OMPVV_INFOMSG test_target_defaultmap.F90:42] test_defaultmap_on
[OMPVV_INFOMSG test_target_defaultmap.F90:112] test_defaultmap_off
[OMPVV_ERROR test_target_defaultmap.F90:194]  Condition scalar_char /= 'a' failed
[OMPVV_ERROR test_target_defaultmap.F90:19]  Condition test_defaultmap_off() .ne. 0 failed
[OMPVV_INFOMSG ompvv.F90:231] The value of errors is 2
[OMPVV_RESULT test_target_defaultmap.F90] Test failed on the device.
```

**Comment by dmcdougall** [comment]
Now I have to figure out what the test is doing :)

**Comment by frederik-h** [comment]
> @frederik-h does it mean you suggest a different route? Or do you mean this is still a problem that GCC has to work on?

I am sorry that I do not have the time to look into this in more detail right now. I just wanted to see if the test failure is related to some recent GCC changes which, apparently, is not the case.

In contrast to what I wrote above, the C test also fails with host execution:
```
[OMPVV_INFO: test_target_defaultmap.c:89] Test is running on host.
[OMPVV_INFO: test_target_defaultmap.c:43] test_defaultmap_off
[OMPVV_ERROR: test_target_defaultmap.c:70]  Condition scalar_char != 'b' failed 
[OMPVV_ERROR: test_target_defaultmap.c:71]  Condition scalar_short != 20 failed 
[OMPVV_ERROR: test_target_defaultmap.c:72]  Condition scalar_int != 33 failed 
[OMPVV_ERROR: test_target_defaultmap.c:73]  Condition scalar_float != 6.5f failed 
[OMPVV_ERROR: test_target_defaultmap.c:74]  Condition scalar_double != 20.45 failed 
[OMPVV_ERROR: test_target_defaultmap.c:75]  Condition scalar_enum != VAL4 failed 
[OMPVV_ERROR: test_target_defaultmap.c:92]  Condition test_defaultmap_off() failed 
[OMPVV_INFO: test_target_defaultmap.c:94] The value of errors is 6.
[OMPVV_RESULT: test_target_defaultmap.c] Test failed on the host.
```

**Comment by frederik-h** [comment]
If you compare GCC's behavior with the description of the implicit data mapping rules from the OpenMP 5 specification, Section 2.19.7 "Data-Mapping Attribute Rules, Clauses, and Directives", it seems that GCC is doing the right thing here. That is, it copies the scalar variables and hence the change from the `omp target` block is not visible after leaving the block. At the bottom of the section, the spec says [sic]:
> If none of the above rules applies then a scalar variable is not mapped, but instead has an implicit
> data-sharing attribute of mapped, but instead has an implicit data-sharing attribute of firstprivate
> (see Section 2.19.1.1 on page 270)

Although this sentence is somewhat broken, when comparing with OpenMP 4.5, Section 2.15.5 on page 216, it is clear that the second alternative (firstprivate) is the correct one:

> If a defaultmap(tofrom:scalar) clause is not present then a scalar variable is not
> mapped, but instead has an implicit data-sharing attribute of firstprivate

This implies that the value should be copied and this seems to be what GCC does.

**Comment by tob2** [comment]
Some additional remarks:
- **Regarding `kind=16`:** gfortran supports `real(kind=16)` (i.e. 128bit floating-point numbers) and, hence, `complex(kindx=16)` – at least on some systems, including x86-64 (when `libquadmath` support is enabled [= default]). But `complex(8)` (64 bit reals/double precision) makes more sense – especially for offloading. _However, I want to point out that standard does not indicate any relation between the `kind` value and its meaning; NAG f95 uses, e.g., 1,2,3 as values. Fortran 2008 provides the kind constants (parameters) `REAL32`, `REAL64`, `REAL128` in the `ISO_FORTRAN_ENV` module, which permits `complex(kind=real64)`. This is more readable, fully standard conform and might be more portable, iff the compiler supports this bit of  F2008._
- **gfortran**'s OpenMP support is lagging a bit behind (compared with gcc/g++), hence, OpenMP 4.5 **'`defaultmap`'** is not yet supported, see [PR86416](https://gcc.gnu.org/PR86416)
- Still, GCC (gcc, g++, gfortran)'s **implicit mapping** is correct according to OpenMP 4.5 (2.15.5) and 5.0 (cf. 2.19.7, last two bullet points; gfortran: except for OMP 5's handling of allocatable/pointer scalars, cf. [PR92568](https://gcc.gnu.org/PR92568)) _[Link to OpenMP 5 spec](https://www.openmp.org/spec-html/5.0/openmpsu109.html#x142-6180002.19.7), scroll down to last bullet point before "2.19.7.1 map Clause"_
- As written in the item before and in the two previous comments, **for scalars, implicit mapping is '`firstprivate`',** which implies that the three '`defaultmap`' test cases should also get for the host that value changes in the target section do not influence the value after that section. _(Cf. OpenMP 4.5, "2.15.3.4 `firstprivate` Clause": [C/C++] "For variables of non-array type, the initialization occurs by copy assignment")_

Therefore, without `defaultmap(tofrom:…)` – a scalar modified in a target section will keep its (previous) value outside/after the target section due to the '`firstprivate`'. Thus,
- `tests/4.5/target/test_target_defaultmap.F90`: In `test_defaultmap_off`, the assumption that the value changes if `isHost = omp_is_initial_device()` is .true. does not make sense; OpenMP implies that it shall be firstprivate and hence  unmodified after the 'target' block.
- `tests/4.5/target/test_target_defaultmap.F90`: Likewise for `test_defaultmap_off`.
- `tests/4.5/target_teams_distribute_parallel_for/test_target_teams_distribute_parallel_for_defaultmap.c`: Likewise for `test_defaultmap_off`

**Comment by josemonsalve2** [comment]
@tob2  Thank you very much really valuable insights. I will fix these firstprivate errors, you are completely right. 

From what you said, it seems like kind=16 is still valid for our tests, if we are assuming Fortran 2008 support. But it seems like OpenMP 4.5 does not support 2008. The latest support seems to be 2003 (according to section 1.6)

**Comment by tob2** [comment]
@josemonsalve2 wrote:
> From what you said, it seems like kind=16 is still valid for our tests, if we are assuming Fortran 2008 support. But it seems like OpenMP 4.5 does not support 2008.

Well, whether a specific compiler/hardware combination supports "real128" (the data type, not the constant) is independent of Fortran 2008; it usually will have either the value 16 or to -1 (= no such type available). gfortran additionally has "10" (80-bit double-extended precision, the x87 coprocessor precision). — I think many compilers only support real32 + real64 and neither the 80 nor the 128 bit version, even if the hardware does.

All available kind numbers are in the REAL_KINDS array of the same ISO_FORTRAN_ENV module. — Whether an offload target supports all of them, is another question. It probably only supports REAL32 [often: = 4] (single) and REAL64 [often: = 8] (double precision) – unless the 'target' section is running on the host itself.

I (@tob2) wrote:
>  **gfortran**'s OpenMP support is lagging a bit behind (compared with gcc/g++), hence, OpenMP 4.5 **'`defaultmap`'** is not yet supported, see [PR86416](https://gcc.gnu.org/PR86416)

Correction: True in general but '`defaultmap(tofrom:scalar)`' is supported (= only defaultmap possible in OpenMP 4.5); just the OpenMP 5 extensions aren't; I mixed that up when trying myself. _(I still have to read the PR beyond just the subject.)_

In terms of issues, gfortran currently maps nonarray character strings as 'tofrom' and not as firstprivate, cf.  [PR92920](https://gcc.gnu.org/PR92920). _This needs to be fixed after confirming that's indeed what the spec states and/or the spec writers intended._ ;-)

**Comment by tob2** [comment]
> In terms of issues, gfortran currently maps nonarray character strings as 'tofrom' and not as firstprivate, cf. [PR92920](https://gcc.gnu.org/PR92920).

Turns out that GCC correctly handles them – see quote below. That means that `tests/4.5/target/test_target_defaultmap.F90` has to be modified such that it check for `character` not as if it is a scalar with `firstprivate` but a nonscalar with `map(tofrom:`.

OpenMP 4.5 + 5.0: "1.2.6 Data Terminology"):
> ***scalar variable***
>  For C/C++: A scalar variable, as defined by the base language.
>  For Fortran: A scalar variable with intrinsic type, as defined by the
>               base language, **excluding character type.**"

**Comment by richardbleikamp** [comment]
Just a quick comment that "tob2"s description of real types in Fortran 90, F2003, F2008, and F2018 is correct.  Including that REAL*16 (whatever KIND it is) is NOT required in a standard conforming compiler, nor COMPLEX*32 a,b,c nor COMPLEX(KIND=16) a,b,c.

There are programmers that recommend the use of SELECTED_REAL_KIND as the most portable way to specify KIND numbers and get what you really want.
i.e., for the flang compiler running on an Intel or AMD CPU host:
~/test$ flang slk.f90 
~/test$ ./a.out
 real kind with at least 5 digits of precision=            4
 real kind with at least 10 digits of precision=            8
~/test$ cat slk.f90
	real (kind=selected_real_kind(p=5)) a
	real (kind=selected_real_kind(p=10)) b
	write(6,*) "real kind with at least 5 digits of precision=", kind(a)
	write(6,*) "real kind with at least 10 digits of precision=", kind(b)
	end

note that fortran doesn't actually require a real datatype with any specific precision be available in any implementation, just that REAL and DOUBLE PRECISION exist, with some unspecified KIND number, and DOUBLE PRECISION must have more precision that REAL.  But an 8 bit and 16 bit REAL type (obviously with somewhat useless precision and exponent ranges) would satisfy the standard.

So for benchmarking on the host CPU, selected_real_kind, with P=5 and P=10 are going to be the most portable mechanism available, and will work as expected on all 32/64 bit systems.  GPU support is not really standardized per se, but the same mechanism should work with most compilers.
See https://gcc.gnu.org/onlinedocs/gfortran/SELECTED_005fREAL_005fKIND.html for some examples.
I'd observe that using just plain REAL and DOUBLE PRECISION declarations is also completely portable.  and REAL (KIND=4) and REAL (KIND=8), although not "guaranteed" to be portable by any standard, are also very portable.  

So, I suggest that the COMPLEX(KIND=16) declarations be replaced by COMPLEX (KIND=8),
so they are double precision complex, or use the COMPLEX(KIND=selected_real_kind(p=10)) declaration style.

**Comment by josemonsalve2** [comment]
If I leave the default precision and do not specify a KIND, that should also be ok, right? I have added you as PR reviewer. 

Thanks for all that info @tob2, really valuable.

**Comment by richardbleikamp** [comment]
Correct.  Using "COMPLEX x,y,x" and "DOUBLE COMPLEX xx,yy,zz" is portable, always available, and most likely what you really want.

**Comment by josemonsalve2** [comment]
This issue was solved by PR #44 . Feel free to re-open if necessary
